Safari: audio elements with event listeners are not getting garbage collected

jernoble opened the pull request:
#### a2ddc404d28da5ea719bef25ead5d58e7a609ff0
<pre>
Safari: audio elements with event listeners are not getting garbage collected
<a href="https://bugs.webkit.org/show_bug.cgi?id=262485">https://bugs.webkit.org/show_bug.cgi?id=262485</a>
rdar://116347723

Reviewed by Youenn Fablet.

Clarify the implementation of HTMLMediaElement::virtualHasPendingActivity() to explicitly
handle the cases when an element may have eventListeners but will never fire an event
without its state being mutated through JavaScript.

* LayoutTests/media/media-garbage-collection-expected.txt: Added.
* LayoutTests/media/media-garbage-collection.html: Added.
* LayoutTests/media/video-test.js:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::virtualHasPendingActivity const):
(WebCore::HTMLMediaElement::hasLiveSource const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:

Canonical link: <a href="https://commits.webkit.org/269165@main">https://commits.webkit.org/269165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b1db631b8c990d5c6bf0efa0994852309fa2eba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21744 "2 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23611 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20132 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21985 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22277 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21972 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/21575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18837 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24465 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/18757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19689 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19784 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19902 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23844 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/20383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17361 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19505 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5191 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23936 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20305 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6385,17 +6385,47 @@ void HTMLMediaElement::resume()
+        return true;
+
+    // * It is playing, and is audible to the user:
+    if (isPlaying() && canProduceAudio())
```
Can audio restart automatically when doing a swipe back using the b/f cache?
If so, canProduceAudio might return false at one point but true later on without JS interaction.

jernoble replied:
...or, I guess, if it was interrupted due to being backgrounded, another process began playing audio, etc. I'll handle this case.

youennf commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6385,17 +6385,47 @@ void HTMLMediaElement::resume()
+            return true;
+    }
+
+    // Otherwise, the media element will fire no events at event listeners, and
```
What about the case of being paused because outside of the visible viewport?
If user scrolls and media element gets visible again, it might restart playing without JS interaction.

jernoble replied:
If the element is in the DOM, none of the contents of this method matter, because the DOM will prevent the element from being collected.

youennf reviewed (approved):
LGTM, there might be corner cases though (see possible cases below, in which case we need to fix these) but hopefully we will find them all.

eric-carlson commented on LayoutTests/media/media-garbage-collection.html:
```diff
@@ -0,0 +1,127 @@
+
+            waitFor(event) { return waitFor(this.video, event); }
+            addEventListener(event, callback) { return this.video.addEventListener(event, callback); } 
+        };
```
Nice!

jernoble committed:
Safari: audio elements with event listeners are not getting garbage collected
https://bugs.webkit.org/show_bug.cgi?id=262485
rdar://116347723

Reviewed by Youenn Fablet.

Clarify the implementation of HTMLMediaElement::virtualHasPendingActivity() to explicitly
handle the cases when an element may have eventListeners but will never fire an event
without its state being mutated through JavaScript.

* LayoutTests/media/media-garbage-collection-expected.txt: Added.
* LayoutTests/media/media-garbage-collection.html: Added.
* LayoutTests/media/video-test.js:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::virtualHasPendingActivity const):
(WebCore::HTMLMediaElement::hasLiveSource const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:

Canonical link: https://commits.webkit.org/269165@main
